File: n-queens.java
public class Solution {
    public ArrayList<String[]> solveNQueens(int n) {
        this.n = n;
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                board[i][j] = '.';
        ans = new ArrayList<String[]>();
        col = new boolean[n];
        diag1 = new boolean[n + n - 1];
        diag2 = new boolean[n + n - 1];
        search(board, 0);
        return ans;
    }

    private void search(char[][] board, int r) {
        if (r == n) {
            String[] rows = new String[n];
            for (int i = 0; i < n; i++)
                rows[i] = new String(board[i]);
            ans.add(rows);
            return;
        }
        for (int c = 0; c < n; c++) {
            if (col[c] || diag1[n - 1 - r + c] || diag2[r + c]) continue;
            col[c] = diag1[n - 1 - r + c] = diag2[r + c] = true;
            board[r][c] = 'Q';
            search(board, r + 1);
            board[r][c] = '.';
            col[c] = diag1[n - 1 - r + c] = diag2[r + c] = false;
        }
    }

    private int n;
    private ArrayList<String[]> ans;
    private boolean[] col;
    private boolean[] diag1;
    private boolean[] diag2;
}

File: n-queens.cc
class Solution {
public:
    vector<vector<string> > solveNQueens(int n) {
        this->n = n;
        vector<bool> col(n);
        vector<bool> diag1(n + n -1);
        vector<bool> diag2(n + n - 1);
        vector<int> states;
        ans.clear();
        search(0, states, col, diag1, diag2);
        return ans;
    }
    
    void search(int row, vector<int>& states, vector<bool>& col,
            vector<bool>& diag1, vector<bool>& diag2) {
        if (row == n) {
            ans.push_back(vector<string>(n, string(n, '.')));
            for (int i = 0; i < n; i++)
                ans.back()[i][states[i]] = 'Q';
            return;
        }
        for (int i = 0; i < n; i++) {
            if (col[i] || diag1[(n - 1) - row + i] || diag2[i + row])
                continue;
            states.push_back(i);
            col[i] = diag1[(n - 1) - row + i] = diag2[i + row] = true;
            search(row + 1, states, col, diag1, diag2);
            states.pop_back();
            col[i] = diag1[(n - 1) - row + i] = diag2[i + row] = false;            
        }
    }
private:
    int n;
    vector<vector<string> > ans;
};
